File: go/internal/domain/quantity.go
package domain

import "fmt"

// QuantityMax is the shared inclusive maximum for a strictly positive quantity.
const QuantityMax int64 = 2147483647

// Quantity is a strictly positive number of stock units.
type Quantity int64

// NewQuantity validates that the amount is strictly positive and within the
// shared maximum.
func NewQuantity(value int64) (Quantity, error) {
	if value <= 0 {
		return 0, fmt.Errorf(
			"new quantity: %w",
			InvalidOrder{Reason: fmt.Sprintf("quantity must be strictly positive, got %d", value)},
		)
	}
	if value > QuantityMax {
		return 0, fmt.Errorf(
			"new quantity: %w",
			InvalidOrder{Reason: fmt.Sprintf("quantity exceeds %d, got %d", QuantityMax, value)},
		)
	}
	return Quantity(value), nil
}

// MustQuantity is NewQuantity for infallible literals; it panics on violation
// and is intended for tests and wiring code with compile-time-known inputs.
func MustQuantity(value int64) Quantity {
	q, err := NewQuantity(value)
	if err != nil {
		panic(err)
	}
	return q
}

File: rust/crates/domain/src/quantity.rs
//! Quantity value object: a strictly positive integer.

use std::fmt;

use crate::error::InvalidOrder;

/// Shared quantity maximum (`i32::MAX`) so every language pack can represent it.
pub const QUANTITY_MAX: u32 = 2_147_483_647;

/// An amount of stock that must be strictly positive.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Quantity {
    value: u32,
}

impl Quantity {
    /// Validates and constructs a quantity.
    ///
    /// # Errors
    ///
    /// Returns [`InvalidOrder`] when `value` is zero or above [`QUANTITY_MAX`];
    /// the type cannot hold negatives at all.
    pub fn new(value: u32) -> Result<Self, InvalidOrder> {
        if value == 0 {
            return Err(InvalidOrder::new(
                "quantity must be strictly positive, got 0",
            ));
        }
        if value > QUANTITY_MAX {
            return Err(InvalidOrder::new(format!(
                "quantity exceeds {QUANTITY_MAX}, got {value}"
            )));
        }
        Ok(Self { value })
    }

    /// Returns the strictly positive amount.
    #[must_use]
    pub fn value(self) -> u32 {
        self.value
    }
}

impl fmt::Display for Quantity {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.value)
    }
}

#[cfg(test)]
mod tests {
    // Documented test-side suppressions (see src/order.rs).
    #![allow(
        clippy::unwrap_used,
        reason = "tests build known-valid values and read errors"
    )]

    use super::*;

    #[test]
    fn zero_is_rejected() {
        let error = Quantity::new(0).unwrap_err();
        assert!(error.reason().contains("strictly positive"));
    }

    #[test]
    fn positive_values_are_accepted_and_preserved() {
        assert_eq!(Quantity::new(1).unwrap().value(), 1);
        assert_eq!(Quantity::new(QUANTITY_MAX).unwrap().value(), QUANTITY_MAX);
    }

    #[test]
    fn above_max_is_rejected() {
        let error = Quantity::new(QUANTITY_MAX + 1).unwrap_err();
        assert!(error.reason().contains("exceeds"));
    }

    #[test]
    fn display_renders_the_value() {
        assert_eq!(Quantity::new(7).unwrap().to_string(), "7");
    }
}
